addons/mail_group/models/mail_group.py

```diff
@@ -0,0 +1,319 @@
+        values.update({
+            'author_id': author_id,
+            'email_from': email_from,
+            'model': self._name,
+            'res_id': self.id,
+            'partner_ids': [],
+        })
+        attachments = kwargs.get('attachments') or []
+        attachment_ids = kwargs.get('attachment_ids') or []
+        attachement_values = Mailthread._message_post_process_attachments(attachments, attachment_ids, values)
+        values.update(attachement_values)
+        mail_message = Mailthread._message_create(values)
+
+        # Create the group message associated
+        group_message = self.env['mail.group.message'].create({
+            'mail_group_id': self.id,
+            'mail_message_id': mail_message.id,
+        })
+
+        # Check the moderation rule to determine if we should accept or reject the email
+        email_normalized = email_normalize(kwargs.get('email_from'))
+        moderation = self.env['mail.group.moderation'].search([
+            ('mail_group_id', '=', self.id),
+            ('email', '=', email_normalized),
+        ], limit=1)
+
+        if not self.moderation or (moderation and moderation.status == 'allow'):
+            group_message.action_accept()
+
+        elif moderation and moderation.status == 'ban':
+            group_message.action_discard()
+
+        elif self.moderation and self.moderation_notify:
+            self.env['mail.mail'].sudo().create({
+                'author_id': self.env.user.partner_id.id,
+                'email_from': self.env.user.company_id.catchall_formatted or self.env.user.company_id.email_formatted,
+                'body_html': group_message.mail_group_id.moderation_notify_msg,
+                'subject': 'Re: %s' % (kwargs.get('subject', '')),
+                'email_to': kwargs.get('email_from'),
+                'auto_delete': True,
+                'state': 'outgoing'
+            })
+
+    # ------------------------------------------------------------
+    # MODERATION
+    # ------------------------------------------------------------
+
+    def _join_group(self, email, partner_id=None):
+        self.ensure_one()
+
+        if partner_id:
+            partner = self.env['res.partner'].browse(partner_id).exists()
+            email = partner.email
+
+        email_normalized = email_normalize(email)
+
+        existing_member = self.member_ids.filtered_domain([('email_normalized', '=', email_normalized)])
+        if existing_member:
+            raise UserError(_('This email is already registered in this group.'))
+
+        member = self.env['mail.group.member'].create({
+            'partner_id': partner_id,
+            'email': email,
+            'mail_group_id': self.id,
+        })
+
+        if self.moderation_guidelines:
+            # Automatically send the guidelines to the new member
+            self.action_send_guidelines(member)
+
+    def _leave_group(self, email):
```

Comment:
Should probably support partner or email to give an API easier to use.

Reply:
I added a method `_get_member` to standardize to way we retrieve member